apps/aecore/test/aec_tx_pool_tests.erl

```diff
@@ -0,0 +1,155 @@
+             %% Start `aec_keys` merely for generating realistic test
+             %% signed txs - as a node would do.
+             TmpKeysDir = mktempd(),
+             ok = application:ensure_started(crypto),
+             {ok, _} = aec_keys:start_link(["mypassword", TmpKeysDir]),
+             TmpKeysDir
+     end,
+     fun(TmpKeysDir) ->
+             ok = aec_keys:stop(),
+             ok = application:stop(crypto),
+             {ok, KeyFiles} = file:list_dir(TmpKeysDir),
+             %% Expect two filenames - private and public keys.
+             [_KF1, _KF2] = KeyFiles,
+             lists:foreach(
+               fun(F) ->
+                       AbsF = filename:absname_join(TmpKeysDir, F),
+                       {ok, _} = {file:delete(AbsF), {F, AbsF}}
+               end,
+               KeyFiles),
+             ok = file:del_dir(TmpKeysDir),
+             ok = aec_tx_pool:stop()
+     end,
+     [{"No txs in mempool",
+       fun() ->
+               ?assertEqual({ok, []}, aec_tx_pool:peek(1)),
+               ?assertEqual({ok, []}, aec_tx_pool:peek(3)),
+
+               STx1 = a_signed_tx(<<"pk1">>, 1, 1),
+               ?assertEqual(ok, aec_tx_pool:delete(STx1))
+       end},
+      {"As a healthy network peer, the node stores in mempool txs received from peers and serves txs in mempool to peers",
+       fun() ->
+               %% No txs to serve to peers.
+               ?assertEqual({ok, []}, aec_tx_pool:peek(1)),
+
+               %% Tx received from a peer.
+               STx1 = a_signed_tx(<<"pk1">>, 1, 1),
+               ?assertEqual(ok, aec_tx_pool:push(STx1)),
+
+               %% One tx to serve to peers.
+               ?assertEqual({ok, [STx1]}, aec_tx_pool:peek(1)),
+
+               %% Other tx received from a peer.
+               STx2 = a_signed_tx(<<"pk2">>, 1, 2),
+               ?assertEqual(ok, aec_tx_pool:push(STx2)),
+
+               %% Two tx2 to serve to peers.
+               ?assertEqual({ok, [STx2]}, aec_tx_pool:peek(1))
+       end},
+      {"Mempool follows longest chain",
+       fun() ->
+               %% Prepare a few txs.
+               STx1 = a_signed_tx(<<"pk1">>, 1, 1),
+               STx2 = a_signed_tx(<<"pk2">>, 1, 1),
+               STx3 = a_signed_tx(<<"pk3">>, 1, 1),
+               STx4 = a_signed_tx(<<"pk4">>, 1, 1),
+               STx5 = a_signed_tx(<<"pk5">>, 1, 1),
+
+               %% Some txs received from peers.
+               ?assertEqual(ok, aec_tx_pool:push(STx2)),
+               ?assertEqual(ok, aec_tx_pool:push(STx4)),
+               ?assertMatch({ok, [_,_]}, aec_tx_pool:peek(10)),
+
+               %% A block inserted in chain.
+               ?assertEqual(ok, aec_tx_pool:delete(STx1)),
+               ?assertEqual(ok, aec_tx_pool:delete(STx2)),
+               ?assertEqual({ok, [STx4]}, aec_tx_pool:peek(10)),
+
+               %% A tx already in chain now received from peers.
+               ?assertEqual(ok, aec_tx_pool:push(STx2)),
```

Comment:
If `STx2` is already in a chain and we re-add it to the pull, such tx will be removed from a pool in mining attempt? Mining shall recognize it as duplicate?

Reply:
> If STx2 is already in a chain and we re-add it to the pool, such tx will be removed from the pool in mining attempt?

Txs are removed from pool when added in chain - see GH-236. Mining itself is not meant to remove txs from pool. After successful mining, if mined block added to chain then txs in block shall be then removed.

> Mining shall recognize it as duplicate then?

Yes, thanks to nonce. Mining attempts to apply txs. Each tx is meant to  contain initiating account and its nonce, and each tx is meant to ensure nonce in account is strictly less than nonce in txs. So duplicated tx is detected by mining, and mining shall delete such tx from pool.